### ISSUE TITLE
RFC: Add a "step" axis for logging metrics in MLflow

### ISSUE THREAD

**Issue by dbczumar** [mlflow/mlflow#1129]
**Motivation / Overview**

Training machine learning models is a process of iterative improvement. Models are successively exposed to sections of training data and adjusted based on their performance. Therefore, ML developers often need to observe how model performance changes throughout the training process. There are multiple ways to quantify the progress of the training process; here are several:

* The amount of time, in seconds (wall clock time), that the training process has been running
* The number of times that the model has been exposed to the entire training dataset. This is the number of training epochs.
    * Models often make multiple passes through the training dataset.
* The number of batches (fixed-size subsets of training data) to which the model has been exposed. This is the number of training iterations.

MLflow currently provides an API for tracking model performance metrics; however, MLflow metrics can only be specified as a function of wall clock time. To paraphrase, wall clock time is the only supported x-axis on which metrics can recorded and visualized; wall clock time is the only supported x-coordinate for a metric entry.

The purpose of this RFC is to introduce support for recording and visualizing performance metrics against different types of axes, enabling ML developers to answer questions such as:

* “What was the training accuracy of my model after the 50th training epoch?”
* “What was the precision of my model when evaluating the test data set after the 20th training iteration?”

**Summary of Proposed Solution**
At a high level, we propose to introduce a `step` attribute to MLflow metrics via the following high-level tasks:
* Introduce a step attribute to MLflow’s Metric entity definition
* Update backend store definitions to handle the new attribute
* Update the metric write path to create Metric entities that supply values for the step attribute
    * This includes client APIs, protos, the backend store write path, etc.
* Expand plotting capabilities of the MLflow UI to allow the user to toggle between the step axis and the “wall clock time” axis.

For more information, please see the full RFC below.

**RFC link**
The full RFC is available as a Google Document at: https://docs.google.com/document/d/17yHR_xOvoEJQBT-D-y4QGfc3OvHqP-eGWwfMcGBIH2k/edit?usp=sharing

Please leave comments and feedback within Google Docs!

**Comment by jeremyjordan** [comment]
related issue: #32 
i'd recommend talking with @smurching

**Comment by dbczumar** [comment]
Closing this RFC because the features it describes have been implemented.
